design/kubebuilder-migration.md

```diff
@@ -0,0 +1,64 @@
+This will make extending Contour with additional CRDs as we need to more straightforward.
+
+## Goals
+
+- Migrate Contour to use Kubebuilder [controller-runtime][1] instead of informers.
+
+## Non Goals
+
+- Change anything after the DAG builder
+- Implement the new service-api types as part of this proposal
+
+## Background
+
+Two years ago, @davecheney built Contour to use custom-wired code to talk to Kubernetes.
+At the time, this was a sensible decision, as there were no well-developed frameworks or common methods for interacting with Kubernetes.
+Since then, the community has done a large amount of work to simplify the construction of controllers, currently embodied in the kubebuilder [controller-runtime][1] library.
+
+This has become an issue now because we are looking at adding support for further CRDs, (the SIG-Network service-apis) and want to leave the door open for more if we need to.
+
+In addition, moving to using controller-runtime means that we have access to frameworks for helping with:
+
+- leader election
+- defaulting values for CRDs
+- webhooks to translate between CRD versions (related to #2068)
+
+without having to custom-build that code.
+
+Lastly, this will allow us to consume the new service-api types and any other CRDs (including our own), without generating the client-go code to make Informers.
+The service-api project currently does not generate these.
+
+This is a pretty big refactor though, so we should agree it's a good idea before starting.
+
+## High-Level Design
+
+I propose a multi-phase approach that can be broadly summarized as:
+
+- Move things that use custom Kubernetes code to use more standard implementations (particularly Caches, but there may be others). This will involve a few changes at least.
+- Move `contour serve` to using the controller-runtime framework for its serving.
+
+In particular, the controller-runtime framework requires Controllers to have a Reconcile function that is supplied the details of an object that has changed, and the controller is expected to perform whatever actions it needs to to ensure that the new object correctly reflects the state of the world.
+It's a reconciliation loop in a function, in other words.
+
+Currently Contour uses an Eventhandler with Add, Update and Delete methods for each type it supports for this function, also with an added Cache that only keeps things that the Contour DAG should care about.
+
+### Moving towards more standard Kubernetes interactions
+
+In order to do this, we will need to move all of the functions that require access to cached Kubernetes objects to use a more standard Kubernetes cache (instead of the custom Contour one.)
+
+Once that is done, we'll need to move the code that performs changes from Add, Update and Delete, into a singular Reconcile function that figures out what has changed and does that.
+
+The detailed design section is omitted because this work will of necessity need to change as it goes.
+
+## Alternatives Considered
```

Comment:
Another alternative is that we could use client-go dynamic client which doesn't require the generated client like we do today and I suspect it's what the controller-runtime is using.

Reply:
That's a good point, I will check that out.